[2024-09-12 22:04:51,240] [INFO] [real_accelerator.py:158:get_accelerator] Setting ds_accelerator to cuda (auto detect)
/workspace/Oobleck/oobleck/__init__.py
node_index: 4
[2024-09-12 22:04:54,380] [WARNING] [agent.py:154:_launch_workers] Profile data for model gpt2 not found. Exception Error parsing json file: /workspace/Oobleck/tmp/profiles/gpt3_1_3B-16-8-1/mb16.json Launching profiler...
[2024-09-12 22:04:56,166] [INFO] [real_accelerator.py:158:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2024-09-12 22:04:59,349] [INFO] [profiler.py:270:profile] Profiling model gpt2. rank 4, world_size 8, num_workers_per_node 1.
Dataset: model_name: gpt2, dataset_path: wikitext, dataset_name: wikitext-2-raw-v1, max_seq_length: 1024
[2024-09-12 22:05:01,254] [INFO] [profiler.py:276:profile] Dataset loaded
model_name: gpt2, model_tag: gpt3_1_3B
[2024-09-12 22:05:02,187] [INFO] [profiler.py:284:profile] model loaded
[2024-09-12 22:05:24,836] [INFO] [profiler.py:292:profile] init pg. master_addr: 172.21.0.42, master_port: 23456
[2024-09-12 22:05:26,276] [INFO] [profiler.py:306:profile] Profiling model execution latency.
[2024-09-12 22:05:26,276] [INFO] [profiler.py:112:profile_execution_layers] before barrier
[2024-09-12 22:05:36,823] [INFO] [profiler.py:114:profile_execution_layers] after barrier
[2024-09-12 22:05:36,824] [INFO] [profiler.py:120:profile_execution_layers] before broadcast
[2024-09-12 22:05:36,824] [INFO] [profiler.py:122:profile_execution_layers] after broadcast
[2024-09-12 22:05:36,827] [INFO] [profiler.py:317:profile] Profiling cross-node allreduce latency.
[2024-09-12 22:05:36,827] [INFO] [profiler.py:160:profile_allreduce_across_nodes] Profile allreduce across 8 nodes latency
[2024-09-12 22:05:37,015] [INFO] [profiler.py:326:profile] Profiling in-node allreduce latency.
[2024-09-12 22:05:37,016] [INFO] [profiler.py:209:profile_allreduce_in_node] Profile allreduce within a node latency
[2024-09-12 22:05:38,092] [INFO] [agent.py:189:_launch_workers] in agent. my_ip_port 172.21.0.46:2220, node_ip_ports ['172.21.0.42:2220', '172.21.0.42:2221', '172.21.0.42:2222', '172.21.0.42:2223', '172.21.0.46:2220', '172.21.0.46:2221', '172.21.0.46:2222', '172.21.0.46:2223']
[2024-09-12 22:05:38,092] [INFO] [agent.py:192:_launch_workers] Launching worker 0...
[2024-09-12 22:05:40,030] [INFO] [real_accelerator.py:158:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2024-09-12 22:05:43,304] [INFO] [worker.py:23:worker_main] Initializing Oobleck Engine...
[2024-09-12 22:05:43,304] [INFO] [worker.py:24:worker_main] in worker main: my_ip 172.21.0.46:2220
/root/miniconda3/envs/oobleck/lib/python3.10/site-packages/transformers/training_args.py:1281: FutureWarning: using `no_cuda` is deprecated and will be removed in version 5.0 of ðŸ¤— Transformers. Use `use_cpu` instead
  warnings.warn(
Dataset: model_name: gpt2, dataset_path: wikitext, dataset_name: wikitext-2-raw-v1, max_seq_length: 1024
[2024-09-12 22:05:45,292] [INFO] [engine.py:635:_initialize_engine] model arguments: {'n_embd': 2048, 'n_head': 32, 'n_positions': 1024, 'num_hidden_layers': 24}
model_name: gpt2, model_tag: gpt3_1_3B
mb size: 26, ar across: 26, ar in: 26
Loading done. creating layer execution results...
Returning from get_profiler_results
[2024-09-12 22:05:46,458] [INFO] [engine.py:677:_initialize_engine] Number of nodes range: (2, 8)
min_nodes: 2, max_nodes: 8, gpus: 1, 
Create pipeline templates
StageExecutionResult[0:13] with 1 devices
StageExecutionResult[14:25] with 1 devices
StageExecutionResult[0:9] with 1 devices
StageExecutionResult[10:17] with 1 devices
StageExecutionResult[18:25] with 1 devices
StageExecutionResult[0:7] with 1 devices
StageExecutionResult[8:14] with 1 devices
StageExecutionResult[15:20] with 1 devices
StageExecutionResult[21:25] with 1 devices
StageExecutionResult[0:6] with 1 devices
StageExecutionResult[7:11] with 1 devices
StageExecutionResult[12:16] with 1 devices
StageExecutionResult[17:21] with 1 devices
StageExecutionResult[22:25] with 1 devices
StageExecutionResult[0:5] with 1 devices
StageExecutionResult[6:10] with 1 devices
StageExecutionResult[11:14] with 1 devices
StageExecutionResult[15:18] with 1 devices
StageExecutionResult[19:22] with 1 devices
StageExecutionResult[23:25] with 1 devices
[2024-09-12 22:05:47,013] [DEBUG] [agent.py:298:on_receive_response] Receiving: (<Response.FORWARD_RANK0_PORT: 5>, <RequestType.UNDEFINED: 0>)
[2024-09-12 22:05:47,013] [DEBUG] [agent.py:243:on_receive_worker_port] agent recv TCP Store port 36507 from master
StageExecutionResult[0:4] with 1 devices
StageExecutionResult[5:8] with 1 devices
StageExecutionResult[9:12] with 1 devices
StageExecutionResult[13:15] with 1 devices
StageExecutionResult[16:19] with 1 devices
StageExecutionResult[20:22] with 1 devices
StageExecutionResult[23:25] with 1 devices
StageExecutionResult[0:4] with 1 devices
StageExecutionResult[5:8] with 1 devices
StageExecutionResult[9:11] with 1 devices
StageExecutionResult[12:14] with 1 devices
StageExecutionResult[15:17] with 1 devices
StageExecutionResult[18:20] with 1 devices
StageExecutionResult[21:23] with 1 devices
StageExecutionResult[24:25] with 1 devices
[2024-09-12 22:05:47,235] [INFO] [worker.py:27:worker_main] Initializing torch.distributed...
[2024-09-12 22:05:47,235] [DEBUG] [engine.py:725:initialize_distributed] regenerate rank_map {'172.21.0.42:2220': [0], '172.21.0.42:2221': [1], '172.21.0.42:2222': [2], '172.21.0.42:2223': [3], '172.21.0.46:2220': [4], '172.21.0.46:2221': [5], '172.21.0.46:2222': [6], '172.21.0.46:2223': [7]}
[2024-09-12 22:05:47,235] [INFO] [engine.py:739:initialize_distributed] init pg: rank 4, world_size: 8, rank_map: {'172.21.0.42:2220': [0], '172.21.0.42:2221': [1], '172.21.0.42:2222': [2], '172.21.0.42:2223': [3], '172.21.0.46:2220': [4], '172.21.0.46:2221': [5], '172.21.0.46:2222': [6], '172.21.0.46:2223': [7]}
[2024-09-12 22:05:47,235] [INFO] [engine.py:755:initialize_distributed] Waiting for a port information...
[2024-09-12 22:05:47,235] [INFO] [engine.py:758:initialize_distributed] Received torch master: 172.21.0.42:2220.36507
[2024-09-12 22:05:47,257] [INFO] [comm.py:637:init_distributed] cdb=None
[2024-09-12 22:05:47,258] [INFO] [engine.py:775:initialize_distributed] [rank: 4] Distributed initialization is done.
[2024-09-12 22:05:47,258] [INFO] [worker.py:34:worker_main] Instantiating pipelines...
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 22:05:47,680] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 22:05:47,915] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 22:05:48,252] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 22:05:48,477] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
[2024-09-12 22:05:48,567] [INFO] [instantiator.py:256:get_best_execution_plan] Best execution plan: 4 x <oobleck.PipelineTemplate.2nodes> pipelines (b: 16)
B: 64
dataloader: mbs 16
[2024-09-12 22:06:04,050] [INFO] [engine.py:159:_reconfiguration_listener_fn] ReconfigureEngine: start reconfigure listening
[2024-09-12 22:06:04,051] [INFO] [worker.py:36:worker_main] Begin training...
step: 0
step in try: 0
load_mbs: 0.03597688674926758, forward time: 10.420616626739502, backward time: 25.605220317840576
load_mbs cnt: 16, forward cnt: 16, backward cnt: 16
training step: 0 done
/root/miniconda3/envs/oobleck/lib/python3.10/site-packages/torch/cuda/memory.py:416: FutureWarning: torch.cuda.memory_cached has been renamed to torch.cuda.memory_reserved
  warnings.warn(
/root/miniconda3/envs/oobleck/lib/python3.10/site-packages/torch/cuda/memory.py:424: FutureWarning: torch.cuda.max_memory_cached has been renamed to torch.cuda.max_memory_reserved
  warnings.warn(
step: 1
step in try: 1
load_mbs: 0.02586197853088379, forward time: 9.941499710083008, backward time: 25.626473903656006
load_mbs cnt: 16, forward cnt: 16, backward cnt: 16
training step: 1 done
step: 2
step in try: 2
load_mbs: 0.02478933334350586, forward time: 9.948907852172852, backward time: 25.578975915908813
load_mbs cnt: 16, forward cnt: 16, backward cnt: 16
training step: 2 done
step: 3
step in try: 3
load_mbs: 0.024452924728393555, forward time: 9.95396614074707, backward time: 25.597426652908325
load_mbs cnt: 16, forward cnt: 16, backward cnt: 16
training step: 3 done
step: 4
step in try: 4
load_mbs: 0.02413034439086914, forward time: 9.954956769943237, backward time: 25.602431774139404
load_mbs cnt: 16, forward cnt: 16, backward cnt: 16
training step: 4 done
step: 5
step in try: 5
load_mbs: 0.02124500274658203, forward time: 9.954524278640747, backward time: 25.596688985824585
load_mbs cnt: 16, forward cnt: 16, backward cnt: 16
training step: 5 done
step: 6
step in try: 6
load_mbs: 0.0233001708984375, forward time: 9.948444604873657, backward time: 25.59391140937805
load_mbs cnt: 16, forward cnt: 16, backward cnt: 16
training step: 6 done
step: 7
step in try: 7
load_mbs: 0.02028369903564453, forward time: 9.957728147506714, backward time: 25.572368383407593
load_mbs cnt: 16, forward cnt: 16, backward cnt: 16
training step: 7 done
step: 8
step in try: 8
load_mbs: 0.023589134216308594, forward time: 9.946537256240845, backward time: 25.585243463516235
load_mbs cnt: 16, forward cnt: 16, backward cnt: 16
training step: 8 done
step: 9
step in try: 9
load_mbs: 0.021197080612182617, forward time: 9.938034296035767, backward time: 25.58252191543579
load_mbs cnt: 16, forward cnt: 16, backward cnt: 16
training step: 9 done
[2024-09-12 22:12:33,554] [INFO] [engine.py:932:train] Training is done. Waiting for synchronization...
[2024-09-12 22:12:34,039] [INFO] [engine.py:935:train] Training done.
